src/init-action.ts

```
@@ -143,9 +143,10 @@ async function run() {
     repositoryNwo,
     logger
   );
-  void featureFlags.preloadFeatureFlags();
 
   try {
+    await featureFlags.preloadFeatureFlags();
```

Can we still void this?  There's no point in preloading the flags if we void them.  Alternatively can we kick this promise off at the start of the `try` and await it at the end?

---

If we void it, then we need to explicitly handle any errors. This may be fine if the security events error should only be a warning. Maybe `Encountered an error while trying to load feature flags` should be a warning as well?

If either of these are warnings, then we will try to grab feature flags again later. Should we also set the feature flags to all falsy if we fail to grab them?

---

There's a comment in there about why we can't do that — in brief some feature flags affect alerts, so transient errors with the API endpoint could lead to alert churn if we don't crash.

---

OK...so, we can't do it for non-403 errors, but we can for the case that we are trying to solve. 

So...my suggestion is for 403 errors, we can return falsy values, but for other errors, we must end the workflow run.

---

> So...my suggestion is for 403 errors, we can return falsy values, but for other errors, we must end the workflow run.

Yep, that sounds sensible to me.  I edited my comment above with a suggestion about telemetry, but I don't think it blocks the PR.

---

> One of the reasons to be within the try though would be error telemetry. To get this, we could change the preload method to tryPreloadFeatureFlags and have it fail silently. The main disadvantage would probably be a few extra API requests in the event that the requests error with non 403s, which isn't the end of the world.

Sorry...a little confused here. The comment says that `ml_powered_queries_enabled` means we must fail hard if we can't get the feature flag. So, for non-403s, we would just get a single API failure and the run would end, right?

---

I think if we void the `preloadFeatureFlags` promise (as we do in `main`), then a hard fail exception won't get captured by the `catch` and so we won't send an "init aborted" status report.  We have a couple of hard fail exceptions that are currently logged by that telemetry, and I was just thinking about how we could make `Encountered an error while trying to load feature flags` one of them.

---

Based on our thread above, I think we should be throwing all errors from the function above. And as you suggest, we can also run the function at the start of the try block, and `await` it at the end.

---

Wait...this won't work. We are `await`ing the promise, only after it thows. We are going to have to add a rejection handler, capture the error, and then optionally throw it later. It will have to look something like this:

```js
async function errorFunction() {
  throw new Error()
}

async function wait() {
  return new Promise(res => setTimeout(res, 1000))
}

(async function() {
  try {
    let maybeError
    const promise = errorFunction().catch(e => maybeError = e)
    await wait()
    await promise

    if (maybeError) {
      throw maybeError
    }
  } catch (e) {
    console.log('error is caught')
  }

})()
```

---

Hmm, I'm not sure the complexity of that is worth the slight perf improvement from preloading the feature flags — perhaps we should remove the preloading.

---

I think that would be fine. I don't think the API request is particularly long running...is it?

---

No, I'm not concerned.